src/greenlight/step.clj

```diff
@@ -299,38 +301,48 @@
                             ::reports @reports)]
     (binding [ctest/report (partial swap! reports conj)
               *pending-cleanups* (atom [])]
-      (try
-        (let [test-fn (::test step)
-              timeout (::timeout step 60)
-              inputs (collect-inputs system ctx step)
-              step-future (future (test-fn inputs))
-              output (deref step-future (* 1000 timeout) ::timeout)]
-          (if (= output ::timeout)
-            (do
-              (future-cancel step-future)
-              [(output-step
-                 :timeout
-                 (format "Step timed out after %d seconds" timeout))
-               ctx])
-            (let [report-types (group-by assert/report->outcome @reports)
-                  passed? (and (empty? (::assert/fail report-types))
-                               (empty? (::assert/error report-types)))]
-              [(output-step
-                 (if passed? :pass :fail)
-                 (->> report-types
-                      (map #(format "%d %s"
-                                    (count (val %))
-                                    (name (key %))))
-                      (str/join ", ")
-                      (format "%d assertions (%s)"
-                              (count @reports))))
-               (save-output step ctx output)])))
-        (catch Exception ex
-          (let [message (format "Unhandled %s: %s"
-                                (.getSimpleName (class ex))
-                                (.getMessage ex))]
+      (let [test-fn (::test step)
+            timeout (::timeout step 60)
+            inputs (collect-inputs system ctx step)
+            step-future (future (test-fn inputs))
+            result (try
+                     (deref step-future (* 1000 timeout) ::timeout)
+                     (catch ExecutionException ex
+                       (ex-cause ex)))]
```

Comment:
Does this need to catch both `Exception` and `ExecutionException`? The old implementation is catching `Exception`. But maybe the `deref` can only throw `ExecutionException`?

Reply:
Yeah, from some quick testing I think this should catch Exceptions as well. Deref on a Clojure future seems to behave similarly to java.util.concurrent.Future#get(), which [declares that it throws the following](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html#get--):


> Throws:
    [CancellationException](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CancellationException.html) - if the computation was cancelled
    [ExecutionException](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutionException.html) - if the computation threw an exception
    [InterruptedException](https://docs.oracle.com/javase/8/docs/api/java/lang/InterruptedException.html) - if the current thread was interrupted while waiting